### PR TITLE
feat: Create Public DID (VDRI & REST support)

### LINKED ISSUE #845: VDRI support & REST endpoint for creating public DIDs
Add create DID feature in VDRI http binding and also add REST endpoint to create public DIDs

### PR DESCRIPTION
**Pull request by sudeshrshetty**
- add rest endpoint `/vdri/create-public-did` to created public DIDs
using VDRI of agent
- VDRI Build() for httpbinding to create public DIDs using custom request
builders
- `/vdri/create-public-did` to accept optional `request header` parameter to add custom headers to 
request being sent to http binding endpoint.
- closes #845

Signed-off-by: sudesh.shetty <sudesh.shetty@securekey.com>

### CODE REVIEW

**Comment by troyronda** [comment]
This doesn’t seem like general path but rather vdri?

File: pkg/restapi/operation/common/operation.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+	b, err := json.Marshal(args)
+	if err != nil {
+		return err
+	}
+
+	return json.Unmarshal(b, v)
+}
+
+// buildSideTreeRequest request builder for sidetree public DID creation
+func buildSideTreeRequest(didDoc *did.Doc) (io.Reader, error) {
```
Comment (fqutishat) [inline pkg/restapi/operation/common/operation.go]: aries code should not know about sidetree

**Comment by sudeshrshetty** [comment]
> This doesn’t seem like general path but rather vdri?
> 
> Is this an endpoint in aca-py?

Updated, aca-py doesn't have it.

File: pkg/restapi/operation/common/operation.go (hunk 1 above)
Comment (sudeshrshetty) [reply pkg/restapi/operation/common/operation.go]: Removed, I just added feature to inject optional header to http binding request. Using that header to inject side tree specific minimal headers, and it is working. 
The current create request in REST API follows `{header:xxy, payload:xxz}` pattern. We can modify it anytime in future to add support to other public DIDs.

File: pkg/framework/aries/api/vdri/vdri.go [hunk 2]
```diff
@@ -113,6 +118,14 @@ func WithServiceEndpoint(serviceEndpoint string) DocOpts {
 	}
 }
 
+// WithRequestBuilder allows to supply request builder
+// which can be used to add headers to request stream to be sent to HTTP binding URL
+func WithRequestBuilder(builder func(payload []byte) (io.Reader, error)) DocOpts {
+	return func(opts *CreateDIDOpts) {
+		opts.RequestBuilder = builder
+	}
+}
```
Comment (llorllale) [inline pkg/framework/aries/api/vdri/vdri.go]: I'm wondering if this option is even needed at all. I would have assumed the specific VDRI implementation would take care of building the request and serializing whatever needs to be serialized. The VDRI implementation for the DID method can be picked with the `VDRI.Accept(method string)` method.

File: pkg/vdri/httpbinding/vdri.go [hunk 3]
```diff
@@ -57,12 +65,103 @@ func (v *VDRI) Accept(method string) bool {
...
+	resDoc, err := v.sendCreateRequest(reqBody)
+	if err != nil {
+		return nil, fmt.Errorf("failed to send create DID request: %s", err)
+	}
+
+	return resDoc, nil
+}
+
+func (v *VDRI) sendCreateRequest(req io.Reader) (*did.Doc, error) {
+	httpReq, err := http.NewRequest(http.MethodPost, v.endpointURL, req)
```
Comment (llorllale) [inline pkg/vdri/httpbinding/vdri.go]: please leave a TODO for configuring timeouts properly on these external calls

File: pkg/restapi/operation/common/models.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+// CreatePublicDIDResponse model
+//
+// This is used for returning public DID created
+//
+// swagger:response createPublicDIDResponse
+type CreatePublicDIDResponse struct {
+
+	// in: body
+	DID *did.Doc `json:"did"`
```
Comment (llorllale) [inline pkg/restapi/operation/common/models.go]: I believe it's better to return the base64-encoded raw bytes of the DID doc for the reasons laid out in hyperledger/aries-rfcs#243

File: pkg/vdri/httpbinding/vdri.go [hunk 5]
```diff
@@ -57,12 +65,103 @@ func (v *VDRI) Accept(method string) bool {
...
+	}
+
+	var reqBody io.Reader
+	if docOpts.RequestBuilder != nil {
+		reqBody, err = docOpts.RequestBuilder(docBytes)
+		if err != nil {
+			return nil, fmt.Errorf("failed to build request : %s", err)
+		}
+	} else {
+		bytes.NewReader(docBytes)
```
[the remark this answers is not in the document]
Comment (sudeshrshetty) [reply pkg/vdri/httpbinding/vdri.go]: done

File: pkg/restapi/operation/common/operation.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+		return
+	}
+
+	if request.CreatePublicDIDParams == nil || request.Method == "" {
+		errors.SendHTTPBadRequest(rw, InvalidRequestErrorCode, fmt.Errorf("invalid method name"))
+		return
+	}
+
+	logger.Debugf("creating public DID for method[%s]", request.Method)
+	fmt.Println("creating public DID for method[]", request.Method)
```
[the remark this answers is not in the document]
Comment (sudeshrshetty) [reply pkg/restapi/operation/common/operation.go]: done

File: pkg/framework/aries/api/vdri/vdri.go (hunk 2 above)
Comment (sudeshrshetty) [reply pkg/framework/aries/api/vdri/vdri.go]: we thought about the same when we started doing this change. SDK has feature of injecting custom VDRI implementation through opts/context. It cannot be done with controller APIs.

But this option can be used to modify request being sent to http binding URL.  REST API use this option to add http binding endpoint specific headers.

File: pkg/vdri/httpbinding/vdri.go (hunk 3 above)
Comment (sudeshrshetty) [reply pkg/vdri/httpbinding/vdri.go]: Agreed, will be addressed in #855 , added a TODO

File: pkg/restapi/operation/common/models.go (hunk 4 above)
Comment (sudeshrshetty) [reply pkg/restapi/operation/common/models.go]: Agreed, will be addressed in #855 , added a TODO

File: pkg/framework/aries/api/vdri/vdri.go (hunk 2 above)
Comment (llorllale) [reply pkg/framework/aries/api/vdri/vdri.go]: @sudeshrshetty 

> It cannot be done with controller APIs.

Sounds like some kind of refactor is needed to support this use case?

Comment (sudeshrshetty) [reply pkg/framework/aries/api/vdri/vdri.go]: This is what this PR is about. HTTP Binding implementation creates DID with minimal features like public keys and service endpoint (from current agent) and submits it to http binding endpoint URL. This option will help us to add custom header, in long run we have to go as you suggested.

This PR is to unblock other public DID related features in controller API

Comment (sudeshrshetty) [reply pkg/framework/aries/api/vdri/vdri.go]: Will be handled as part of next PR for #860

**Pull request merged**